File: Engine/src/main/java/org/etieskrill/engine/graphics/model/Node.java
package org.etieskrill.engine.graphics.model;

import lombok.Getter;
import lombok.ToString;
import org.etieskrill.engine.Disposable;
import org.etieskrill.engine.entity.component.Transform;
import org.etieskrill.engine.entity.component.TransformC;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

@Getter
@ToString(exclude = {"parent", "children"})
public final class Node implements Disposable {

    private final String name;
    private final TransformC transform;
    private final @Nullable Node parent;
    private final List<Node> children;
    private final List<Mesh> meshes;
    private final @Nullable Bone bone;

    public Node(String name, @Nullable Node parent, TransformC transform, List<Mesh> meshes, @Nullable Bone bone) {
        this.name = name;
        this.transform = transform;
        this.parent = parent;
        this.children = new ArrayList<>();
        this.meshes = meshes;
        this.bone = bone;
    }

    public Transform getHierarchyTransform() {
        return getHierarchyTransform(0);
    }

    public Transform getHierarchyTransform(int numNodesIgnoredFromRoot) { //TODO precompute
        Node root = this;
        Stack<Node> line = new Stack<>();
        line.push(this);
        while ((root = root.getParent()) != null) //Traverse up the tree to the root, recording every node
            line.push(root);

        for (int i = 0; i < numNodesIgnoredFromRoot; i++) line.pop();

        Transform transform = new Transform();
        while (!line.isEmpty()) //Traverse down the tree, applying every transform in succession
            transform.apply(line.pop().getTransform());

        return transform;
    }

    @Override
    public void dispose() {
        meshes.forEach(Mesh::dispose);
        children.forEach(Node::dispose);
    }

}

File: Engine/src/commonMain/kotlin/org/etieskrill/engine/graphics/model/Node.kt
package org.etieskrill.engine.graphics.model

import org.etieskrill.engine.entity.component.Transform
import org.etieskrill.engine.entity.component.TransformC
import org.etieskrill.engine.entity.component.applyInternal
import org.joml.times

data class Node(
    val name: String,
    val transform: TransformC = Transform(),
    val parent: Node? = null,
    val children: MutableList<Node> = mutableListOf(),
    val meshes: List<Mesh> = listOf(),
    val bone: Bone? = null
) {

    fun getGlobalTransform(numNodesIgnoredFromRoot: Int = 0): Transform { //TODO precompute
        var root: Node? = this
        val nodes = mutableListOf<Node>()
        while (root != null) {
            nodes += root
            root = root.parent
        }

        nodes.dropLast(numNodesIgnoredFromRoot)

        val transform = Transform()
        nodes.asReversed().forEach { transform.apply(it.transform) }

        return transform
    }

    override fun toString() = "Node{name='$name', transform=$transform, meshes=$meshes, bone=$bone}"

}
